src/toil/job.py

```diff
@@ -3026,3 +3030,32 @@ def convertPromises(kwargs):
                 foundPromisedRequirement = True
         return foundPromisedRequirement
 
+
+class JobStoreFile(str):
+    """
+    A class to wrap the job store file id returned by writeGlobalFile and any attributes we may want
+    to add to it.
+
+    It is easier to implement a startswith method here, than to inherit string because that makes
+    pickling and unpickling really hard.
+    """
+    # str isa  new-style class and we should modify __new__ instead of __init__
+    @classmethod
+    def __new__(cls, *args, **kwargs):
+        # args becomes (<class>, fileStoreID, filePath)
+        obj = str.__new__(cls, args[1])
+        _, obj.fileStoreID, obj.filePath = args
+        assert isinstance(obj.fileStoreID, (str, unicode))
+        if obj.filePath is not None:
+            # This is the case when we are unpickling.  At this point, size is already in __dict__,
+            # moreover it probably doesn't even exist anymore.
+            assert isinstance(obj.filePath, (str, unicode))
+            assert os.path.exists(obj.filePath)
```

Comment:
Also, pull this out into factory method `JobStoreFile.forPath(path)` such that `__new__` doesn't mix all these concerns. `__new__` takes care of construction and unpickling, `__reduce__` takes care of pickling.